Implement basic support for vectorized instructions.

delcypher opened the pull request:
We use LLVM's Scalarizer pass to remove most vectorized code so that the
Executor only needs to support the InsertElement and ExtractElement
instructions.

This pass was not available in LLVM 3.4 so to support that LLVM version
the pass has been back ported.

There is no support for LLVM 2.9.

There are a few limitations to this implementation.

* The InsertElement and ExtractElement index cannot be symbolic.
* Due to the lack of proper floating point support the
`float_vector_constant_llvm*.ll` tests currently do not pass.

delcypher commented:
@andreamattavelli @ccadar This is my attempt at upstreaming the vector instruction support I have in my fork of KLEE.

I don't think the current lack of support for a symbolic index for the InsertElement/ExtractElement instructions is a problem. The vectorized code likely comes from LLVM's optimizers where those indices will be concrete. During my testing I never came across symbolic indices.

Hopefully with this we can kill LLVM 2.9.

delcypher commented:
@ccadar @andreamattavelli @MartinNowack One thing I wanted your opinion on is type checking the instructions.

If the `Scalarizer` pass screws up we might end up with vector operands being unexpectedly given to instructions.

So we could add these checks to many instructions in `Executor::executeInstruction()`

```c++
    assert(!(ki->inst->getOperand(0)->getType()->isVectorTy()) &&
           "Invalid operand type");
    assert(!(ki->inst->getOperand(1)->getType()->isVectorTy()) &&
           "Invalid operand type");
```

or even tighter (e.g. for the Add instruction)

```
    assert(ki->inst->getOperand(0)->getType()->isIntegerTy() &&
           "Invalid operand type");
    assert(ki->inst->getOperand(1)->getType()->isIntegerTy() &&
           "Invalid operand type");
```

The problem with this is this will likely decrease the performance of `assert` builds.

So my current PR doesn't add any of these assertions. However we could add these, if you value correctness over performance.

ccadar commented:
Thanks, Dan, that's an important contribution.  Regarding your last point: what happens now if the Scalarizer pass misses an instruction?  Don't you hit an existing assert?

delcypher commented:
All tests are passing apart from [macOS which is flaky](https://travis-ci.org/klee/klee/jobs/235698695). The failure appears unrelated.

ccadar commented:
The failure seems unrelated to flaky tests. 
More importantly, it's not a good idea to disregard test failures -- as we've seen in the past, it's easy to misidentify the root cause.

delcypher commented:
@ccadar 
> Thanks, Dan, that's an important contribution. Regarding your last point: what happens now if the Scalarizer pass misses an instruction? Don't you hit an existing assert?

I don't know about missing an instruction but if I don't run the pass the Executor just carries on executing. Only the `shuffle_element.c` test fails because it hits the `ShuffleVector` instruction which is supposed to be removed by the pass.

I don't have tests that apply operators (e.g. Add) to vector data. I should probably add some and see what happens.

andreamattavelli commented on lib/Core/Executor.cpp:
```diff
@@ -333,6 +334,7 @@ namespace klee {
 const char *Executor::TerminateReasonNames[] = {
   [ Abort ] = "abort",
   [ Assert ] = "assert",
+  [ BadVectorAccess] = "bad_vector_access",
```
Wrong formatting.

delcypher replied:
What do you want to see instead? `badvectoraccess`?

andreamattavelli replied:
I was referring to the square braket `]`: 

```cpp
[ BadVectorAccess ] = "bad_vector_access",
```

andreamattavelli commented on test/VectorInstructions/float_vector_constant_llvm34.ll:
```diff
@@ -0,0 +1,244 @@
+; FIXME: Need to upstream floating point support to make this test pass
```
I would remove this test from this PR and add it when floating-point support is upstreamed.

ccadar replied:
I agree.

andreamattavelli commented on test/VectorInstructions/float_vector_constant_llvm35.ll:
```diff
@@ -0,0 +1,244 @@
+; FIXME: Need to upstream floating point support to make this test pass
```
I would remove this test from this PR and add it when floating-point support is upstreamed.

ccadar replied:
+1

andreamattavelli commented on test/VectorInstructions/float_vector_constant_llvm36.ll:
```diff
@@ -0,0 +1,244 @@
+; FIXME: Need to upstream floating point support to make this test pass
```
I would remove this test from this PR and add it when floating-point support is upstreamed.

ccadar replied:
+1

andreamattavelli commented:
@ccadar 
> The failure seems unrelated to flaky tests.

It is: I don't understand why but it seems that Travis-CI is affected by some strange network-related issues. Basically, homebrew fails to download the bottled version of LLVM 3.4 from the repository (which is currently my website).

@delcypher 
> So we could add these checks to many instructions in Executor::executeInstruction()

I am personally in favor of adding more asserts. If we need performance, we compile without assertions (I already do this).

@ccadar @delcypher 
> what happens now if the Scalarizer pass misses an instruction? Don't you hit an existing assert?

I think that by adding more asserts will likely make KLEE fail if an instruction escapes from the pass. Does it make sense to you?

delcypher commented:
@andreamattavelli 
> One major point to address (later) is the support to symbolic in vectorized instructions.

As I mentioned I have doubts about this actually happening in practice. More testing will reveal whether this is actually needed.

delcypher commented:
@andreamattavelli 
> I think that by adding more asserts will likely make KLEE fail if an instruction escapes from the pass. Does it make sense to you?

I've added asserts for all the instructions that I think need them. I've also added a few tests to check that scalarization is happening as expected. I also fixed a bug in the order the passes were being run.

andreamattavelli commented:
@delcypher I'm happy with the current state of this PR. Before approval, I'd like to test it on a few case studies.

andreamattavelli commented:
@delcypher it does work perfectly on all the case studies where I experienced vectorization-related issues.
I'll wait for @ccadar to comment.

delcypher commented:
> @ccadar 
@delcypher I'm happy with the current state of this PR. Before approval, I'd like to test it on a few case studies.

Thanks. Do you want me to squash the commits into a single commit?

andreamattavelli commented:
@delcypher 
> Do you want me to squash the commits into a single commit?

I think it is better to wait for @ccadar's approval, then squash, wait for Travis, and merge.

delcypher commented:
@andreamattavelli I've pushed a few additional changes. Could you re-run on your case studies?

I forgot to add asserts to the select instruction. When I added additional tests it turned out those asserts were wrong. We might get operands of vector type to the select instruction's true and false operands..

andreamattavelli commented:
@delcypher everything works correctly. I guess that I haven't experience such a problem with `select`. I can try to increase optimization level to see if I can spot it.

delcypher commented:
@andreamattavelli 
>  I can try to increase optimization level to see if I can spot it.

`-Oall-the-things` :fire: 

I changed/removed the asserts on the select instruction because they were wrong so I doubt you'll be able to make them fire.

andreamattavelli commented:
@delcypher
> I changed/removed the asserts on the select instruction because they were wrong so I doubt you'll be able to make them fire.

I'm still using the old version. But no, I cannot make the assertion fail even with the maximum level of optimization.

andreamattavelli commented:
@delcypher @ccadar Going to run this PR through KLEE regression infrastructure just in case.

ccadar commented:
It would be great to run this through the larger regression infrastructure, making sure you enable optimizations.  I'll try to look at the code soon too.  And yes, it would be useful to add this as a single commit.  Thanks!

delcypher commented:
@ccadar @andreamattavelli I just had a though there is a slight problem with the implementation in`lib/Module/KModule.cpp`. Currently we run the `Scalarizer` pass early on (before div/overshift checks because they don't know how to deal with vector instructions).

However if `opts.Optimize` was `true` (I think it's `false` by default right now) then vector instructions might introduced again. We have fine grained control over the optimizations we run so either 

* we need to make sure that the `Optimize()` function does not introduce vector instructions 
or
* we can run the `Scalarizer` pass again after optimization.

Also note that I noticed in the past in my fork of KLEE the `Scalarizer` pass sometimes left behind unnecessary instructions, those can be cleaned up by adding `createDeadCodeEliminationPass()` to the passes. We have to be careful here though because this pass might remove code that performs undefined behaviour.

delcypher commented:
@ccadar @andreamattavelli Can we merge this? PRs are starting to appear (e.g. #669) that are going to conflict with this.

At this point this PR has been reasonably well tested and I think we can attack the problem with `--optimize` in the future.

andreamattavelli commented:
> Can we merge this?

Not so sure. In the previous comment you made some interesting points that still need to be addressed.

delcypher commented:
@andreamattavelli Sure they need be addressed but we don't need to do this right now

* `--optimize` is currently off by default
* KLEE's optimization is broken in several ways.
  - Given that it is been off by default for so long running this passes will likely lead to problems (e.g. unsupported intrinsics)
  - KLEE's optimization pass pipeline was written a long time ago and is completely out of date with 
    what modern Clang does. It likely needs a complete re-write. We should be using LLVM's [PassManagerBuilder](http://llvm.org/docs/doxygen/html/classllvm_1_1PassManagerBuilder.html) if we want to construct a standard compiler pass pipeline. If we want to do something different then someone needs to take ownership of KLEE's custom pass pipeline and invest time to figure out what optimizations make sense for symbolic execution.
  - KLEE's optimization also has confusing command line flags (#366)

In summary I don't think we should block this PR due to the broken optimization feature in KLEE.

The assertions I've added will catch violations of the invariant we expect (i.e. arithmetic/comparision instructions don't receive vector operands).

ccadar commented:
Hi @delcypher , I agree and will try to review and merge this in the next couple of weeks (and hopefully sooner).  However, with respect to dropping 2.9, (#318), support for optimizations is the main remaining showstopper (after this PR), as 2.9 is the only version I know where one can still run fully optimized code.   In any case, I agree treating the deprecation of 2.9 as the next big milestone.

delcypher commented:
@ccadar With this PR in place I was able to run optimized (optimized externally by Clang, not by KLEE) code without problems on my benchmarks. For LLVM 3.4 at least this PR is probably enough to be able to run optimized code in most cases. As you see above @andreamattavelli didn't run into any problems with this PR enabled and optimizations enabled.

For newer LLVMs there are still problems but those can be addressed later given that we don't officially support them yet.

If there are actually examples for LLVM 3.4 where KLEE runs into unhandled intrinsics we need benchmarks to confirm and fix these.

Given the above (and because technically we don't support LLVM > 3.4 yet) I think we can start to deprecated LLVM 2.9 once this PR is merged.

delcypher commented:
@ccadar I've created #678 to track the intrinsics that KLEE is known not to support.

ccadar commented:
@andreamattavelli can I confirm that you would be happy to merge this after I review the code?  
You ran this successfully through the bigger testing infrastructure while compiling everything with `-O2`, correct?

@delcypher , good idea to create #678

ccadar commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2576,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+      // by the next element (second left most) and so on. This means
+      // that we have to adjust the index so we read left to right
+      // rather than right to left.
+      unsigned bitOffset = EltBits*(vt->getNumElements() -i -1);
```
Weird spacing `-i -1` -> `- i - 1`.   I would also add space between and after `*` 
Also, why not reuse `ElemCount`?

delcypher replied:
Yes we should reuse `ElemCount` (it should be const too). The formatting needs to be fixed too.

delcypher replied:
Fixed

ccadar commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2576,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+  case Instruction::InsertElement: {
+    InsertElementInst *iei = cast<InsertElementInst>(i);
+    ref<Expr> vec = eval(ki, 0, state).value;
+    ref<Expr> newElt = eval(ki, 1, state).value;
```
Should we add a check that the type of `newElt` is consistent with the type of the vector elements?

delcypher replied:
We can't do this. KLEE's expression language is completely untyped so `newElt` doesn't have a type. We could insert a check that does something like

```
assert(
  dyn_cast<VectorTy>(i->getOperand(0)->getType())->getElementType() == 
  i->getOperand(1)->getType());
```

This seems unnecessary though because I would expect this to hold by construction (i.e. I would expect LLVM's own asserts to enforce this).

ccadar replied:
Expr has a width, which I regard as a type.

MartinNowack replied:
Yes, we should check the width if it is equivalent.

ccadar commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2576,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    for (unsigned i = 0; i < ElemCount; ++i) {
+      // evalConstant() will use ConcatExpr to build vectors with the
+      // zero-th element left most (most significant bits), followed
+      // by the next element (second left most) and so on. This means
```
Minor: 'leftmost' is spelled as a single word

delcypher replied:
I'll fix that.

delcypher replied:
Fixed

ccadar commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2576,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
-    terminateStateOnError(state, "XXX vector instructions unhandled",
-                          Unhandled);
+    llvm_unreachable(
+        "ShuffleVector instruction should have been removed by scalarize pass");
```
Minor: 'by the Scalarizer pass'

delcypher replied:
I'll fix that.

delcypher replied:
Fixed

ccadar commented on test/VectorInstructions/extract_element_out_of_bounds.c:
```diff
@@ -0,0 +1,21 @@
+  printf("f[2]=%u\n", f[2]);
+  printf("f[3]=%u\n", f[3]);
+  // CHECK: extract_element_out_of_bounds.c:[[@LINE+1]]: Out of bounds read when extracting element
+  printf("f[4]=%u\n", f[4]); // Out of bounds
```
I would consolidate this OOB check into the previous test.

delcypher replied:
Please elaborate. I don't know what you mean by "previous test" here.

ccadar replied:
I mean the test `extract_element.c`  I agree that in general the more tests the better, but at the same time it's never a good idea to add more tests than we need to.  In general, I have tried to always group checks, otherwise it would have been easy to end up with ~1000 tests instead of ~200 without increasing the testing quality.

delcypher replied:
It is not practical to incorporate  `extract_element_out_of_bounds.c` into `extract_element.c` without resorting to macros to have two different source files and two separate runs of KLEE. Given that, I think it's simpler just to have two separate tests.

ccadar replied:
Why?  I would just have the in-bounds asserts first, and the out-of-bounds assert at the end.
Alternatively, you could also combine the in-bounds and respectively out-of-bounds cases for insert and extract element.

delcypher replied:
Because I'm using `--exit-on-error` and I want a "correct" and "incorrect" program. I don't want just an "incorrect" program.

ccadar commented on test/VectorInstructions/extract_element_symbolic.c:
```diff
@@ -0,0 +1,21 @@
+// REQUIRES: geq-llvm-3.4
```
Not sure we need this test, really.

ccadar commented on test/VectorInstructions/insert_element_symbolic.c:
```diff
@@ -0,0 +1,22 @@
+// REQUIRES: geq-llvm-3.4
```
Again, I don't think testing unsupported features is important, and we also need to keep our regression tests from growing too much.

delcypher replied:
I'd rather keep these tests. They force someone who actually changes KLEE's behaviour to actually do something about the tests because they will break if support for symbolic insert/extract is added.

Also, testing KLEE is not the bottleneck in CI, building is. Running the tests is pretty fast and KLEE's test coverage if pretty poor, we need more tests, not less.

ccadar commented on test/VectorInstructions/extract_element.c:
```diff
@@ -1,7 +1,7 @@
 // RUN: %llvmgcc %s -emit-llvm -O0 -g -c -o %t1.bc
 // RUN: rm -rf %t.klee-out
-// RUN: %klee --output-dir=%t.klee-out --exit-on-error %t1.bc
+// RUN: %klee --output-dir=%t.klee-out --optimize=false --exit-on-error %t1.bc
```
What happens with `--optimize`?

delcypher replied:
With `--optimize` all the vector instructions get constant folded away, so we need to use `--optimize=false` to stop that from happening. I probably ought to add a comment about this.

ccadar commented on test/VectorInstructions/extract_element.c:
```diff
@@ -1,7 +1,7 @@
 // REQUIRES: geq-llvm-3.4
```
Of course, this should be squashed into the prior commit.

ccadar commented on lib/Core/Executor.cpp:
```diff
@@ -1935,27 +1935,43 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
     // Arithmetic / logical
 
   case Instruction::Add: {
+    assert(i->getOperand(0)->getType()->isIntegerTy() &&
```
Did you measure the performance impact on those, e.g., on a microbenchmark?

delcypher replied:
No I haven't. I'm not convinced this is the right approach. If you want performance you shouldn't be running an `asserts` build. We already have lots of asserts that are expensive to check.

Do you have a particular benchmark in mind?

MartinNowack replied:
I agree with @ccadar and @delcypher ;)
Too many checks, it clutters the code a lot. Can this be improved?

ccadar commented on lib/Core/Executor.cpp:
```diff
@@ -1922,15 +1922,10 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    assert((i->getOperand(1)->getType() == i->getOperand(2)->getType()) &&
+           "true and false operands to Select have mismatching types");
+    // NOTE: We are do not check that operands 1 and 2 are not of vector type
+    // because the scalarizer pass might not remove these.
```
Isn't this a problem?  Why aren't vector instructions removed from Select?

delcypher replied:
Because you can have a select (if-then-else) instruction that picks from two different vector operands. The result of the select might then be consumed by an `insertelement` or `extractelement` instruction. One of the tests I wrote exposed this so I changed the assert.

delcypher commented on lib/Core/Executor.cpp:
```diff
@@ -1102,8 +1104,16 @@ ref<klee::ConstantExpr> Executor::evalConstant(const Constant *c) {
+      ref<Expr> res = ConcatExpr::createN(cv->getNumOperands(), kids);
+      delete[] kids;
+      assert(isa<ConstantExpr>(res) &&
+             "result of constant vector build not a constant");
```
Yes. I'll fix that.

delcypher replied:
Fixed

delcypher commented:
@ccadar I've rebased and squashed the commits.

delcypher commented:
Is there anything left to do here?

ccadar commented:
I think most things have been addressed, thank you.  I would still prefer consolidating some of the test cases, I think 11 new tests for this feature is a bit much, and as I said, the general guidelines have been to prefer a smaller number of tests (currently ~200) instead of breaking them into smaller units (which could have easily reached ~1000).  Would you mind doing this?

Other than that, I'd like to test this quickly on some performance microbenchmarks.

ccadar commented:
@delcypher , just to say that I've run some microbenchmarks and I'm happy with the results, performance-wise.  I'd still prefer to merge some tests, but if you feel strongly against this, I can approve the PR as it is, and I'll try to do this myself later, perhaps as part of a bigger reorganization of the tests.

delcypher commented:
@ccadar 
> I'd still prefer to merge some tests, but if you feel strongly against this, I can approve the PR as it is, and I'll try to do this myself later, perhaps as part of a bigger reorganization of the tests.

I've merged some of the tests. Personally I don't like doing this because it has made the merged test more complicated.

ccadar commented:
Thanks, @delcypher , can you just squash the last commit into the main one?
The last failure seems unrelated to this (some STP build problem?), but I restarted the job, hopefully it was something spurious.

delcypher commented:
@ccadar I've squashed the two commits into one.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -1102,8 +1104,16 @@ ref<klee::ConstantExpr> Executor::evalConstant(const Constant *c) {
       ref<Expr> res = ConcatExpr::createN(kids.size(), kids.data());
       return cast<ConstantExpr>(res);
+    } else if (const ConstantVector *cv = dyn_cast<ConstantVector>(c)) {
+      ref<Expr> *kids = new ref<Expr>[ cv->getNumOperands() ];
```
Don't use C-style new/free allocation, `llvm::SmallVector` is the better choice and gets deallocated if it goes out of scope.
And one line less code ;)

delcypher replied:
Yes this seems like a good idea. The reason the code is written this way is that I inherited this code from KLEE-FP and didn't bother to change it.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2686,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    ref<Expr> newElt = eval(ki, 1, state).value;
+    ref<Expr> idx = eval(ki, 2, state).value;
+
+    if (!isa<ConstantExpr>(idx)) {
```
As you need a `ConstantExpr` anyway later, do a `dyn_cast<>()` and check for `nullptr`.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2686,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    unsigned EltBits = getWidthForLLVMType(vt->getElementType());
+
+    if (iIdx >= vt->getNumElements()) {
+      // Out of bounds write
```
comment is superfluous at that point.

delcypher replied:
If you know the code well yes, if you don't then I don't think it's superfluous.

MartinNowack replied:
Agree.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2686,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    }
+
+    const unsigned elementCount = vt->getNumElements();
+    ref<Expr> *elems = new ref<Expr>[elementCount];
```
Please use `llvm::SmallVector` here as well

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2686,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    break;
+  }
+  case Instruction::ExtractElement: {
+    ExtractElementInst *eei = cast<ExtractElementInst>(i);
```
`const`ify and move to point where you actually need the variable

delcypher replied:
I disagree. I deliberately wrote it this away trying to avoid "the pyramid of doom" by preferring early exit. Unless you mean doing the `dyn_cast<>` outside of the `if` conditional and then checking if the result is `nullptr` to do early exit. In which I would agree that's a nice change because it means doing the LLVM RTTI style type check once instead of twice.

MartinNowack replied:
Doing the check just once was my point. And I absolutely agree with you to prefer an early exit.

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2686,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
+    ref<Expr> vec = eval(ki, 0, state).value;
+    ref<Expr> idx = eval(ki, 1, state).value;
+
+    if (!isa<ConstantExpr>(idx)) {
```
`dyn_cast` and `const`ify

delcypher replied:
See previous comment about "the pyramid of doom"

MartinNowack commented on lib/Core/Executor.cpp:
```diff
@@ -2566,16 +2686,92 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
   case Instruction::ShuffleVector:
-    terminateStateOnError(state, "XXX vector instructions unhandled",
-                          Unhandled);
+    llvm_unreachable(
```
Why not terminate the state? The whole KLEE execution gets killed.
It is similar to the insufficient system model we have, therefore it should be handled the same way.

delcypher replied:
I was doing that because in principle we should never see the `ShuffleVector` instruction. However `terminateStateOnError()` is more robust and will have well defined behaviour in a release build.

MartinNowack commented on lib/Module/KModule.cpp:
```diff
@@ -307,6 +307,15 @@ void KModule::prepare(const Interpreter::ModuleOptions &opts,
+  // other than InsertElementInst and ExtractElementInst.
+  //
+  // NOTE: Must come before division/overshift checks because those passes
+  // don't know how to handle vector instructions.
```
fix comment: "how"

delcypher replied:
Could you elaborate. I'm not sure what you're suggesting. Are you saying `s/how/"how"/`?

MartinNowack replied:
Ignore this. I must have read something else.

delcypher commented:
@MartinNowack 
> Too many checks, it clutters the code a lot. Can this be improved?

I have two ideas

* MACROS. Define some short hand macros that do the assertion checks
* A sanity check LLVM pass. The Executor would run this pass just before execution starts (i.e. the IR has already been modified so we can assume that what the pass sees is what the Executor executes). This pass would check the instructions where we expect the "no vector operand" invariant to hold and abort if the invariant does not hold.

However I don't really have a problem with the asserts. They actually document an invariant we expect to hold at that point in the code. If we move those checks out of the code and somewhere else it becomes harder to see the relationship between the invariant and the place in the code we expect the invariant to hold. The checks are syntactically are quite long though and will be prone to causing lots of breakages if LLVM's API changes. So the macro option might be sensible here.

delcypher commented:
For reference the macro option would look something like

```c
#define ASSERT_OPERAND_IS_SCALAR_INT(I, N) \
  assert(I->getOperand(N)->getType()->isIntegerTy() && "Operand should be scalar integer")
#define ASSERT_OPERAND_IS_SCALAR_FLOAT(I, N) \
  assert(I->getOperand(N)->getType()->isFloatingPointTy() && "Operand should be scalar float")
```

which would then look like

```c++
  case Instruction::Add: {
    ASSERT_OPERAND_IS_SCALAR_INT(i, 0);
    ASSERT_OPERAND_IS_SCALAR_INT(i, 1);
    ref<Expr> left = eval(ki, 0, state).value;
    ref<Expr> right = eval(ki, 1, state).value;
    bindLocal(ki, state, AddExpr::create(left, right));
    break;
  }

```

MartinNowack commented:
@delcypher I like the idea with the additional pass a lot.
I agree with the point that we have assertions to directly communicate the intention, which is a good thing. I'm not a big fan of hiding this stuff in macros. The problem with those assertions are that they are currently on one of the hottest paths of KLEE. Checking those properties for billions of instructions should be avoided.

The good thing about macros would be that we could easily enable them if we have a debug build and otherwise disable them. And the effort to implement them would be less.

ccadar commented:
Yes, the extra pass sounds best, but I'm happy to go with asserts too at this point.  
I agree tests should be commented more.

ccadar commented:
I'd like to see this PR merged soon, but I am becoming more reticent about having these large number of asserts validating each operation, as they clutter the code and also affect the performance of the assert build (which right now is the default!).

What I suggest is the following:
1. address the easy but useful code comments from @MartinNowack 
2. remove the asserts
3. merge
4. open a todo issue to implement the sanity-check LLVM pass (we should do that, but I don't think it's absolutely needed at this point)

What do you think?

delcypher commented:
@ccadar I'd like to implement the pass because it's not too much work. However it'll be a few days before I have time to do this.

delcypher commented:
I've addressed most of @MartinNowack 's comments but please don't merge yet. I still need to write the pass and remove the asserts.

delcypher commented:
@ccadar @MartinNowack The pass is now implemented and assertions removed. This should be ready to go now. Please review, hopefully it's just "bikeshedding" left as the names I've picked could be better.

delcypher commented:
Eurgh damn it. LLVM 2.9 build is broken.

delcypher commented:
Build is now fixed

ccadar commented on lib/Module/KModule.cpp:
```diff
@@ -390,9 +390,17 @@ void KModule::prepare(const Interpreter::ModuleOptions &opts,
+
+  // Enforce the operand type invariants that the Executor expects
+  if (!roascp->checkPassed()) {
+    klee_error("Unexpected vector operands detected");
```
Really minor, but I would drop "vector" from this message, as there might be other issues detected, e.g., different operand types to a select.

delcypher replied:
Perhaps we ought to rename the pass then because the pass name implies it only checks that some operands are of scalar type. Suggestions for a name? `InstructionOperandTypeCheckPass`?

ccadar replied:
Yes, I think that's a better name.  BTW, I would also add a comment upfront that this pass depends on the scalarizer one.

ccadar commented:
Thanks, @delcypher , this looks good to me, modulo squashing some of the commits.

delcypher commented:
> Thanks, @delcypher , this looks good to me, modulo squashing some of the commits.
@ccadar Done. I've kept the commit that removes the asserts and replaces them with a pass as a separate commit because I think it's important we preserve the history of doing this so it's possible for someone to work out why it's implemented this way.

ccadar commented:
Hi @delcypher , I agree we should have two commits there.  Thanks again for your work on this, I'm happy to approve it.  Let's wait just a bit longer before merging it, to give @MartinNowack an opportunity to react.

MartinNowack commented:
Beside the remark about the comments, I'm fine with the changes. Great job!

delcypher commented:
@MartinNowack 
> Beside the remark about the comments, I'm fine with the changes. Great job!
Comments fixed. I'll merge once I see CI has passed.

ccadar commented:
Thanks again for this contribution, @delcypher , merged!  

@andreamattavelli , @delcypher , @MartinNowack I plan to do a new release now (i.e. today/tomorrow), so please don't merge more things until that is completed. Thanks.
